Features/pipeline

yamadayy opened the pull request:
Implementation pipeline.

YoshiyukiYamada committed:
New files for pipeline. Now skelton.

YoshiyukiYamada committed:
Merge branch 'develop' into features/pipeline

YoshiyukiYamada committed:
move src/tools directory to src/telescope_baseline/tools. Import file names are fixed.

YoshiyukiYamada committed:
Class name is fixed in test_downward.py.
Loop of searching child is added in Analysis and Simulation classes in analysis.py and simulation.py.  Also add test codes.
Add has_parent method in SimComponent class (simcomponent.py).
Remove dummy method which is coded in test codes.

YoshiyukiYamada committed:
Class name is changed from OnTheSkyPosition to OnTheSkyPositoins.
Only above class name change is applied to analysis.py and visitor.py.
Add methods (other than accept) to astrometriccatalogue.py and ontheskypositions.py.
New classes CatalogueEntry (catalogue_entry.py).
Tow visit methods are implemented in simulation.py.

YoshiyukiYamada committed:
astrometriccatalogue.py: add documentation.
catalogue_entry.py: add documentation.
ontheskypositions.py: Add class SkyPosition for individual positiondata, and OnTheSkyPosition class hold the array of SkyPosition. Implements methods.
simcomponent.py: Logic of has_parent method is reverted.
stellarimage.py: Implements methods.
simulation.py: implements visit_si meethod.

YoshiyukiYamada committed:
detectorimage.py: implements for Simulation visitor.
simulation.py: implements visit_di
stellarimage.py: add add_position method.

YoshiyukiYamada committed:
detectorimage.py: add 2 Todos

YoshiyukiYamada committed:
Apply PEP8 rule to simulation.py, visitor.py and ontheskyposition.py.
Implement functions which needed in analysis visitor. Methods extract, to_photon_num, construct_e_psf and x_match in stellarimage.py, get_hdu and load methods in detectorimage.py and add these to visit_di and visit_si methods in analysis.py
Add photoutils and tqdm in requirements.txt.

YoshiyukiYamada committed:
analysis.py
fix visit methods. Constructor of DetectorImage requires filename.

astrometriccatalogue.AstrometricCatalogue
fit_func2 method is added. Attribute result is added. Method calculate_ap_parameters() added.

catalogue_entry: PEP8

detectorimage.py
Method get_parent_list is moved to superclass.
add save_fits and get_array methods.

ontheskypositions.py
Fix set_on_the_sky_list method.
Attribute name position_list is changed to sky_positoin
Empty method solve_distortion is added.

simcompoment.py
get_parent_list is moved from child classes.

simulation.py
fix visit methods.

stellarimage.py
Method get_parent_list is moved to superclass.
Fix method world_to_pixel()

add test_analysis and test_simulation

test_downward.py
test_upward.py
DetectorImage class requires filename.

YoshiyukiYamada committed:
Add/fix documentation.

YoshiyukiYamada committed:
Merge branch 'develop' into features/pipeline

YoshiyukiYamada committed:
Use monkeypatch in pytest for avoiding fail test due to random number.

astronasutarou commented on tests/pipeline/test_analysis.py:
```diff
@@ -0,0 +1,118 @@
+from telescope_baseline.tools.pipeline.stellarimage import StellarImage, OnDetectorPosition
+
+
+def test_step1(monkeypatch):
+    """ test for the step from detector image to stellar position on detector coordinate
+
+    Returns:
+    TODO: Error(s) sometimes occurs because the value is generated from random number.
+    """
+    monkeypatch.setattr(np.random, 'rand', lambda : 0.5)
+    monkeypatch.setattr(np.random, 'randn', lambda : 0)
+    monkeypatch.setattr(np.random, 'uniform', lambda x, y, z: np.zeros(z))
+    a = [OnDetectorPosition(1, 50., 50., 3000.)]
+    w = WCS(naxis=2)
```
`monkypatch` で乱数を返す関数を固定値を返す関数で上書きしていると理解しました．テストのために乱数の seed を固定するのではうまくいかないのでしょうか？乱数を使っている箇所 (`make_image`) をチラ見しましたが固定値で置き換えてしまうと，関数の性質が変わってしまうように感じました．

yamadayy replied:
おっしゃる通り、これがちゃんとしたテストになっているかどうかは甚だ疑問です。ただ、seedを固定するにしても、固定のある程度の長さの乱数列を返す関数に置き換えるにしても、今後ソフトウエアをより現実的にしてゆく過程で乱数を使う箇所が増えるでしょうから、その際に呼ばれる順番が変わって異なる乱数が入ってしまうと、これまでvalidだったものがvalidでなくなる恐れがあります。それはそれで混乱のもとであると考えました。testからvisitorを経由してオブジェクトの中で乱数を呼ぶので、特定の場所の乱数コールだけを固定値にするというのはできるものでしょうか？それができるのであれば、そのように直すのが良いと思いますが。

e-yamane replied:
単体テストと結合テストの違いのお話をした方が良い気がしますね。

> 関数の性質が変わってしまうように感じました．
> おっしゃる通り、これがちゃんとしたテストになっているかどうかは甚だ疑問です。

この辺りもう少し言語化して欲しいです。

> testからvisitorを経由してオブジェクトの中で乱数を呼ぶ特定の場所の乱数コールだけを固定値にするというのはできるものでしょうか？それができるのであれば、そのように直すのが良いと思いますが。

乱数に限らず、別関数、別モジュールの振る舞い影響を受けずに自らのコードの正当性を確認するなら別関数、別モジュールをMockしても良いのですよ。

業務系のシステムの場合は、別のサービスやDBへのアクセスもMockして単体テスト走らせる事が多いです。

yamadayy replied:
Zoomで会話した通りです。(私が何か書かないとunresolved conversationが残ってしまうので)

e-yamane committed:
Test code that continues to maintain what you didn't break by refactoring.

e-yamane committed:
I went out and intensively tested the logic to calculate the positions to add points.

e-yamane committed:
Go out and intensively test the logic to check if the position to add a point is within the DetectorImage

e-yamane committed:
The point totalization must be correct.

e-yamane committed:
Rename the first test method. (Assumed points should be added to the assumed points)

e-yamane committed:
Add pytest-mock

yamadayy committed:
Merge pull request #35 from JASMINE-Mission/features/pipeline_test_detectorImage

Features/pipeline test detector image

YoshiyukiYamada committed:
Remove test_analysis.py and test_simulation.py.
These combined test is moved to each unit test.
Add modules test_astrometriccatalogue.py, test_catalogue_entry.py, test_ontheskyposition.py,and test_stellarimage.py.

astrometriccatalogue.py has two dummy line at the and, and remove one.

YoshiyukiYamada committed:
Merge remote-tracking branch 'origin/features/pipeline' into features/pipeline

YoshiyukiYamada committed:
extract loop from map_to_the_sky to another method.

YoshiyukiYamada committed:
modification of the private member is extracted from the loop

YoshiyukiYamada committed:
extract method from make_image.

YoshiyukiYamada committed:
Number of arguments in fit_astrometry is reduced from 5 to 4. Arrays ls and ty is calculated from Time object within the method.

YoshiyukiYamada committed:
Extract method from world_to_pixel method.

yamadayy commented:
Code Climateの警告はすべて解消しました。昨日教えていただいた方法でテストレポートを書いてみましたが、やはりtest coverageがどこで見えるのかわからなかったので、そちらの評価と対応はできていません。
この状況で問題なければ、山根さんapprove、大澤さんmergeお願いいたします。

YoshiyukiYamada committed:
Remove parenthes

e-yamane reviewed (approved):
実現したいこと自体把握出来てないので、全部は見切れてませんが、detectorImageのmake_imageについては良い感じに出来てると思います！

astronasutarou commented:
追加されたコードが多いのと実現したいことのイメージがまだうまく理解できていないのでもう少しお待ちください……

yamadayy commented:
大澤さん、了解です。Visitorパターンの解説私どこかに書いてslackにあげたと思うのですが、もしVisitorにあまりなじみがなければ、そこを見てもらわないとコードが分かりにくいと思います。

astronasutarou commented:
Slack で該当の文章を見つけました (JASMINE-C2-TN-YY-024_Python_Visitor)．
まずは目を通してみます．

YoshiyukiYamada committed:
Fits file for Test

YoshiyukiYamada committed:
Test file

e-yamane committed:
modeling session

YoshiyukiYamada committed:
c.py is renamed to pipeline.py. Also test_c.py is renamed to test_pipeline.py.  For loading fits file, file path is automatically set from current directory in test_pipeline.py and test_fits_storage.py.

astronasutarou commented on src/telescope_baseline/tools/pipeline/astrometriccatalogue.py:
```diff
@@ -0,0 +1,134 @@
+import astropy.units as u
+
+
+def fit_astrometry(parameter, t, lon, lat):
```
これは fitting の目的関数だと思いますが `fit_astrometry` という名称は confusing かなと思いました．

yamadayy replied:
はい、fitting目的の関数です。名前は考えます。
以前Zoomでお話しした通り、山根さんと相談して設計を大幅に変更しています。pipeline_v2というフォルダーを作って、テストもtests2の下に入れていますが、うまく動くようになったらpipelineとtestsに移すなどの操作をした上で、マージをお願いしようと思っています。
このブランチのマージは少し待ってください。

astronasutarou replied:
> このブランチのマージは少し待ってください

了解しました！
眺めていたら v2 が生まれていたので「あれ？」と思ったところでした．

とりいそぎ resolved としておきます．

e-yamane replied:
> このブランチのマージは少し待ってください。

私は詳しくなかったのでプルリク名に「[WIP]」って付けてわかるようにしてましたが、
この手のプルリクは、Draft Pull Requestにすると良いかもしれませんね。

https://www.publickey1.jp/blog/19/githubwippull_requestdraft_pull_request.html

astronasutarou commented on tests2/test_pipeline.py:
```diff
@@ -0,0 +1,46 @@
+
+
+def get_fits_file_name():
+    file = os.getcwd()
+    sep = file.split('\\')
+    file = ""
+    for i in range(len(sep)):
+        file = file + sep[i] + "\\"
+        if sep[i] == 'telescope_baseline':
+            break
+    file = file + "src\\telescope_baseline\\tools\\pipeline\\for_test.fits"
+    return file
```
file path を扱う場合はライブラリ (pathlib?) を使ったほうがよさそうです．
また，テスト用のデータは `src` 以下に置かないほうがいいかなと思います．

yamadayy replied:
はい、どこが適切ですかね？

astronasutarou replied:
うーむ……おそらく他の場所では使用しなそうですし `tests` ディレクトリの下に置くのが適切だと思います．

astronasutarou commented on tests2/test_pipeline.py:
```diff
@@ -0,0 +1,46 @@
+    return file
+
+
+def test_analysis():
+    c = Pipeline()
+    file = get_fits_file_name()
+    d = DetectorImageCatalogue([
+        FitsStorage.load(file)
+    ])
+    w = WCS(naxis=2)
+    w.wcs.crpix = [256, 256]  # Reference point in pixel
+    w.wcs.cd = [[1.31e-4, 0], [0, 1.31e-4]]  # cd matrix
+    w.wcs.crval = [0, 0]  #
+    w.wcs.ctype = ["GLON-TAN", "GLAT-TAN"]
+    a = c.analysis(d, w)
+    assert(a != None)
```
ファイルが `tests2` にある = テストはしないという意図で良いでしょうか？

astronasutarou commented on tests2/test_fits_storage.py:
```diff
@@ -0,0 +1,9 @@
+from telescope_baseline.tools.pipeline_v2.fitsstorage import FitsStorage
+
+
+def test_load():
+    di = FitsStorage.load(get_fits_file_name())
+    assert(di != None)
```
ファイルが `tests2` にある = テストはしないという意図で良いでしょうか？

e-yamane replied:
これは、私が答えます。tests2を作っているのは、特定のテストクラスを単独で走らせる方法の調査をサボりたかったので、他のテストを走らせることでRed-Green-Refactoringのサイクルが長くなるのを避けるために対象のテストだけ切り出していました。後々testsに統合するイメージです。

yamadayy replied:
いえ、作業中にほかの時間がかかるテストのために待たされたくないので、一時的にtest2に置いています。いろいろ動作が確認できたら、testに移します。

astronasutarou replied:
後に統合予定ということで了解しました．
上の conversation と合わせて resolved とします．

astronasutarou commented:
まだ作業中ということでしたら一度 draft PR に設定変更していただけるとステータスがわかりやすいかもしれません．
…… draft PR でも Code Climate はチェックしてくれる設定になっていたはずです．

yamadayy commented:
混乱させて済みませんでした。draftに変更しておきました。

YoshiyukiYamada committed:
Implements test_pipeline.py.
AstrometricCatalogue: add get_catalogue method.
CatalogueEntry: Change attribute name __id to __stellar_id.
DetectorImage: Use EPSFBuilder and get epsf fitted center.
OnDetectorPosition: Change attribute name __id to __exposure_id.
Pipeline: implements simulation method.
  Add function position_at_certain_time
  change method name from fit_astrometry to lsf_fit_function_for_astrometric_parameter
SkyPosition: Change attribute name from __id to __stellar_id, and add attribute __orbit_id.
StellarImage: Attribute name of SkyPosition is changed.
Add WCSwId class.

YoshiyukiYamada committed:
Method get_fits_file_name is rewritten from with os to with pathlib.

YoshiyukiYamada committed:
extract method of _simulation_stellar_iamge_2_detector_image from simulation.

YoshiyukiYamada committed:
astrometriccataloguebuilder.py, detector_image_builder.py, ontheskyposition_builder.py, stellar_image_builder.py: new class
DetectorImage: Property hdu is added.
OnTheSkyPosition: property orbit_id is added
StellarImage: Argument of type OnDetectorPosition of constructor is changed to the list[OnDetectorPosition].  Window size of extract stellar image is added to its attributes. Also getter of this is added.
Refactoring Pipeline and test_pipeline.

YoshiyukiYamada committed:
AstrometricCatalogueBuilder class is updated.
Add documentation for all classes.

YoshiyukiYamada committed:
Refactored following from CodeClimate warning.

YoshiyukiYamada committed:
Refactored following from CodeClimate warning.

YoshiyukiYamada committed:
Refactored following from CodeClimate warning.

YoshiyukiYamada committed:
Refactored following from CodeClimate warning.

YoshiyukiYamada committed:
Reformat according to PEP8

YoshiyukiYamada committed:
Add DATE-OBS to hdu in detector_image_builder.py
Add tests for builder classes.

YoshiyukiYamada committed:
Changed class names for clarity

YoshiyukiYamada committed:
Changed class names for clarity

YoshiyukiYamada committed:
implements save and load methods.

YoshiyukiYamada committed:
property name is changed.

YoshiyukiYamada committed:
AstrometricCatalogue: add save and load methods
CatalogueEntry: remove unit in pm_ra and pm_dec methods
DetectorImageBuildewr/MapOnDetector: Change property name
Pipeline: change default folder name.w
Add tests and the data necessary for tests.

YoshiyukiYamada committed:
Fix codeclimate error.
Also fix Pycharm warning.

YoshiyukiYamada committed:
Fix assertion error

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue.py:
```diff
@@ -0,0 +1,49 @@
+        file = open(file_name, 'r', newline='')
+        f = csv.reader(file, delimiter=',')
+        for row in f:
+            print(row)
```
printはなくして欲しいです。pythonってロギング機能的なのあるんでしたっけ？

yamadayy replied:
済みません。debug用のものが残ってました。grepしましたが、printはここ以外は例外のところだけです。loggingは考えないといけませんね。

e-yamane replied:
ありがとうございます！

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue.py:
```diff
@@ -0,0 +1,49 @@
+        f = csv.reader(file, delimiter=',')
+        for row in f:
+            print(row)
+            tmp.append(CatalogueEntry(int(row[0]),
+                                      SkyCoord(ra=float(row[1]), dec=float(row[2]), unit=('deg', 'deg'), obstime=row[6],
+                                               pm_ra_cosdec=float(row[3]) * u.mas/u.yr,
+                                               pm_dec=float(row[4]) * u.mas/u.yr, distance=float(row[5])*u.pc,
+                                               frame='icrs'), float(row[7])))
```
saveとloadで処理が非対称なのが気になります。
非対称となっているので、例えば、pm_raと、pm_ra_cosdecが逆変換となっており、知識が流出してます。
難しければ構いませんが、なんとか出来るなら対応いただけるといいなって感じです。

なお、saveとloadが非対称であるが故に、あるAstrometricCatalogueオブジェクトxがあるとして、
x.save('a.csv')
y = AstrometricCatalogue.load('a.csv')
で、xとyが等価であるかが判断付きにくいです。

yamadayy replied:
CatalogueEntryのほうで、本来pm_ra_cosdecとするべきところがpm_raとなっていたので、こちらを修正しました。非対称というのは、引数の順番でしょうか。パラメータの順番をそろえました。

e-yamane replied:
> 非対称というのは、引数の順番でしょうか

save処理がこうなってますが、
```
write.writerow([e.stellar_id, e.ra, e.dec, e.pm_ra_cosdec, e.pm_dec, e.distance, e.coord.obstime, e.mag])
```
であれば、対を成すloadは、
```
tmp.append(CatalogueEntry(
        stellar_id=int(row[0]),
        ra = float(row[1]),
        dec = float(row[2]),
        pm_ra_cosdec = float(row[3]),
        pm_dec = float(row[4])
        distance=float(row[5])
        obstime=row[6]
        float(row[7]))
```
こうなってると、いいなという感じですね。
SkyCoordのプロパティとCatalogueEntryのプロパティの関係式はCatalogueEntryに属する知識なのでCatalogueEntryにカプセル化したいという意図です。

逆を言えば、CatalogueEntryのコンストラクタがIDとSkyCoordを受け取るのであれば、
IDと、SkyCoordを作成するためのプロパティをwriteするのが良いのかも知れません。

どちらが適切かはわかりません。

yamadayy replied:
修正してみましたので、ご確認ください。

e-yamane replied:
ん〜悩ましいですねぇ。saveとloadが対称にはなったのですが、肝心のドメインロジック、
例えばCatalogueEntry.pm_ra_cosdecが、SkyCoord.icrs.pm_ra_cosdec * u.yr / u.mas
であると言うことが漏れ出てこのドメイン知識が３箇所に現れてきてるんですよね。

ここ、csvフォーマットって決まってるんでしたっけ？
ここって SkyCoord -> CatalogueEntry と言う関数従属性があるので、
もし、フォーマットに制約が無く、JSMINEオリジナルフォーマットで良いなら
CSVに出力するデータは、CatalogueEntryのプロパティを保存するという意識を止めて、
SkyCoordのプロパティを出力するのが良いかも知れません。

ここかなりややこしい話してるかも知れません。コメントでやりとりするのは難しそうなのでどこかで話しましょうか？

yamadayy replied:
コメントありがとうございます。フォーマットは決まってないです。今日午後はあいていますので、ご都合が良い時間帯をご指定下さい。いつものパーソナルミーティングでお待ちしております。https://kyoto-u-edu.zoom.us/j/7838600940?pwd=S09EOVhETWZkblNKTEprdWNxNTVMUT09

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,99 @@
+    """ fitting function for leastsq function
+
+    Args:
+        parameter: array for [lon0, lat0, pm_lon_coslat, pm_lat, para]
```
これ、クラスの方が良くないですか？
また、クラスにするのが難しいならarrayよりもtuppleと表現した方が良いかなと思いました。
個人的な感覚かもですが、arrayは、同種のデータが複数あるイメージがあります。
異種のデータをひとまとめにしていると言う意志を強調するならtuppleの方が良い気がします

yamadayy replied:
scipy.optimize.leastsqを呼び出しているので、parameterがnumpy.ndarrayであることは仕様なのです。関数をclass methodでかけるかどうかは試していませんが、クラスにするとしてもクラスの初期化に依存しないようにするとすればstatic methodになるので、とりあえずこのまま放置させてください。

e-yamane replied:
了解です！

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,99 @@
+    para = parameter[4]
+    ls = []
+    ty = []
+    for i in range(len(t)):
+        ls.append(get_sun(t[i]).geocentricmeanecliptic.lon.rad)
+        ty.append(t[i].jyear)
```
添え字でくるくる回すより、

```
for x in t  # xは適切な名前付けて下さい
        ls.append(get_sun(x).geocentricmeanecliptic.lon.rad)
        ty.append(x.jyear)
```
の方が良い気がします。

yamadayy replied:
はい、修正します。

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,99 @@
+    ty = np.array(ty)
+    tc = (np.max(ty) + np.min(ty)) / 2
+    ty = ty - tc
+    lont = np.ndarray((np.size(ty)))
+    latt = np.ndarray((np.size(ty)))
+    residual = np.ndarray((len(ty)))
+    for i in range(len(ty)):
+        lont[i] = lon0 + (para * math.sin(ls[i] - lon0) + pm_lon_coslat * ty[i]) / math.cos(lat0)
+        latt[i] = lat0 + (pm_lat * ty[i] - para * math.sin(lat0) * math.cos(ls[i] - lon0))
+        residual[i] = (lon[i] - lont[i]) ** 2 + (lat[i] - latt[i]) ** 2
```
lontとlattはresidualを計算するための一時退避用の変数っぽいので、配列にする必要は無く、
for分の中で回して良いのでは無いかと思います。

yamadayy replied:
はい、修正します。

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,99 @@
+    return residual
+
+
+class AstrometricCatalogueBuilder:
```
状態持たないなら、変換処理自体はstaticmethodで良いのでは無いですかね？
（その他のBuilderも同様なら同様のご検討を下さい。以降同じコメントはしません）

yamadayy replied:
はい、修正します。

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,99 @@
+        result = []
+        for s in sid:
+            latdata, londata, t = self._time_seriese_of_individual_star(otsp, s)
+            parameter = [np.radians(266), np.radians(-5), 0., 0., np.radians(1 / 3600)]
```
これ(parameter)、値に変化無いのでループの外で定義してもいいと思います

yamadayy replied:
本来はs毎に適切な初期値を計算して入れないといけないところをさぼっているので、TODOをつけました。

e-yamane replied:
了解です

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,99 @@
+        Returns:AstrometriCatalogue
+
+        """
+        sid = []
+        for o in otsp:
+            spl = o.positions_on_the_sky
+            for s in spl:
+                sid.append(s.stellar_id)
+        sid = list(set(sid))
+        result = []
+        for s in sid:
+            latdata, londata, t = self._time_seriese_of_individual_star(otsp, s)
```
position_on_the_skyってどれくらいデータ量ありそうですか？
何気にえげつない回数ループ回るので、最初のループでidだけ取り出すのでは無く、
idをキーにして同一idを持つposition_on_the_skyのlistを持つような辞書を作った方が良いと思いました。

仮にn個position_on_the_skyがあって、idがすべてバラバラだとすると、ここのループは
n + n ** 2
ループが回ると思います。

_add_individual_observationまで考慮するともっと回るかな？すいません。検証できてませんw
とにかくデータが少ないなら良いですが多いのであればベラボーに増えそうだというのが感想です。

yamadayy replied:
10万程度になります。これも、どう最適化するとよいのかすぐには思いつかないので、TODOをつけてそのままで進めたいです。

e-yamane replied:
from_on_the_sky_positionのテストありますか？
あるなら私の方でコード書いてみますが。

yamadayy replied:
test_astrometric_catalogue_builder.pyの中に、test_from_on_the_sky_positionがあります。

e-yamane replied:
プルリクで、別案書いてみました。

https://github.com/JASMINE-Mission/telescope_baseline/pull/57

ただ、それ以前にoptimize.leastsqがとても時間が掛かるようで、これの呼び出し回数が減らないと誤差っぽい気がします。（データ量が増えるとそうでもないかもですが...）
もし、今の方が読みやすいと判断されるのであれば今のままでも良いかもしれません。

yamadayy replied:
#57をマージしました。

e-yamane reviewed (commented):
astrometric_catalogueと、astrometric_catalogue_builderとcatalogue_entry
を一旦見てみました。
多分、構造似てるので似通った指摘が続くことが想定されるので一度ここまでのコメントの是非についてご検討いただけますか？

YoshiyukiYamada committed:
Fix for 1st and 2nd comments

YoshiyukiYamada committed:
Fix for pull request comments

YoshiyukiYamada committed:
loop used in index is changed to object loop if available.

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/detector_image_builder.py:
```diff
@@ -0,0 +1,52 @@
+        for s in range(len(dp)):
+            if dp[s].datetime > t_max:
+                t_max = dp[s].datetime
+            self._generate_a_stellar_image(a, dp, s)  # index s is needed here
```
_generate_a_stellar_imageを見てもdpを配列で、sをdpのindexとして渡す必要性がわかりませんでした。
_generate_a_stellar_imageの中でサボってる処理があって、その時に必要だと言うことであれば、_generate_a_stellar_imageにもtodoコメント付けてくれると助かります。

yamadayy replied:
おっしゃる通りでした。extract methodで分離したmethodの中身を精査すべきでした。直しました。

e-yamane replied:
ありがとうございます

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/detector_image_catalogue.py:
```diff
@@ -0,0 +1,43 @@
+            filename = 'tmp_{:0=4}'.format(i) + '.fits'
+            fpath = Path(p, filename)
+            d.save(str(fpath))
+            i += 1
```
使い勝手的には、saveとloadが逆の方が使いやすい気がします。
つまり、saveの時はfolderが無ければ作って保存、loadの時はfolderが無ければ、読み込み対象が無いので例外。

こっちの方が作りにくいですけどね。

現行のままだとするなら、せめてエラーはNotADirectoryErrorの方が良いと思います。
保存しようとしてるのにFileNotFoundErrorは受け取り側がちょっと「？？？」ってなる気がします。

yamadayy replied:
修正してみましたが、saveの側のテストをどう書くか迷って、TODOにしてしまいました。

e-yamane replied:
ぱっと思いつく物はあるので、私の方で書き足してみましょうかね？
書き足した後に見てもらって、ここのコメントはresolveにしようかと思います。

e-yamane replied:
テスト書き足してみました。問題無ければこのコメントは解決済みにします

e-yamane replied:
Slackでいろいろやりとりしたので解決済みとしますね

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/map_on_detector.py:
```diff
@@ -0,0 +1,51 @@
+        self.__wcs = wcs
+        self.__positions_on_detector = positions_on_detector
+
+    def get_sky_positions(self):
```
あれ？MapOnDetectorがPositionOnDetector -> MapOnDetector の変換ルール知ってるのはNGなのでは？
それぞれの変換はXXXXBuilderが担うのでは？

yamadayy replied:
ご指摘の通りです。MapOnTheSkyBuilderに移しました。まだ少し汚い。

e-yamane replied:
ありがとうございます。MapOnTheSkyBuilderは必要あれば別途コメントします

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/pipeline.py:
```diff
@@ -0,0 +1,109 @@
+from astropy.time import Time
+
+
+class Pipeline:
```
pipelineはsimulationとanalysisはクラス分割しても良いかもしれませんね。
Pipelineのコンストラクタ引数ってsimulation処理でしか使ってませんよね。
analysisだけしたい人から「？？？」ってなりそう

yamadayy replied:
分離しました

e-yamane replied:
ありがとうございます

e-yamane commented on tests/pipeline/test_catalogue_entry.py:
```diff
@@ -0,0 +1,15 @@
+    s = SkyCoord(lon=1., lat=1., unit=('rad', 'rad'), frame='barycentricmeanecliptic', distance=1.0 * u.pc,
+                 pm_lon_coslat=1000 * u.mas / u.yr, pm_lat=500 * u.mas / u.yr)
+    c = CatalogueEntry(1, s, 12.5)
+    assert(abs(c.ra - 15.76) < 0.01)
+    assert(abs(c.dec - 72.34) < 0.01)
+    assert(abs(c.parallax / u.mas - 1000) < 1)
+    assert(abs(c.pm_dec * u.yr / u.mas - 1061) < 1)
+    assert(abs(c.pm_ra * u.yr / u.mas - 351) < 1)
```
浮動小数点の誤差を気にしてるテストならpytest.approxを使うと良いかもしれません。Hw_coeff_test.pyで利用してます

yamadayy replied:
testsのほうは今後使わない予定なので、放置させてください。

e-yamane replied:
了解です

e-yamane reviewed (commented):
第２弾です

YoshiyukiYamada committed:
Fix for symmetry of save and load in astrometric_catalogue.py

YoshiyukiYamada committed:
From unneeded index loop to the object loop.

YoshiyukiYamada committed:
Specification of save and load method in detector_image_catalogue.py is changed. Test for the class is added.

YoshiyukiYamada committed:
get_sky_positions shouuld be the method on xxBuilder.

YoshiyukiYamada committed:
Separate Analysis class and Simulation class from Pipeline class.

e-yamane committed:
Suppress the number of loops in the from_on_the_sky_position function (optimize.leastsq is probably very heavy and does not improve the speed).

YoshiyukiYamada committed:
Remove unused print

yamadayy committed:
Merge pull request #57 from JASMINE-Mission/feature/pipeline_refactor_-from_on_the_sky_position

Suppress the number of loops in the from_on_the_sky_position function…

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/map_on_detector_builder.py:
```diff
@@ -0,0 +1,75 @@
+        for w in wl:
+            if "GLON" not in w.wcs.wcs.ctype[0]:
+                print("Coordinate system " + w.wcs.wcs.ctype[0] + " is not supported")
+                raise ValueError
```
折角なので例外のメッセージとして送出した方が良いと思います。
printだと処理の流れで原因を取りこぼしそうなのと、出来ればprintは実コードからは撲滅したいので

yamadayy replied:
例外メッセージに変更しました。

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/detector_image_catalogue.py:
```diff
@@ -0,0 +1,45 @@
+            filename = 'tmp_{:0=4}'.format(i) + '.fits'
+            fpath = Path(p, filename)
+            d.save(str(fpath))
+            i += 1
```
ここは、indexがファイル名と連動してるので、添え字でのforループがいいかと

yamadayy replied:
変更しました。

e-yamane replied:
ありがとうございます！

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/map_on_the_sky.py:
```diff
@@ -0,0 +1,50 @@
+    """Data holder class of SkyPosition for every orbit.
+
+    """
+    def __init__(self, positions_on_the_sky: list[PositionOnTheSky] = [], orbit_id: int = -1):
```
ココだけじゃ無いかも知れないですが、ここidっぽい変数が後ろにあるのはこのオブジェクトのIDじゃないんですかね？このオブジェクトを一意に識別する目的の物であれば、selfの後ろに定義してあって欲しいです。
ちなみに、引数の順番変更はリファクタリングツールの「Change Signature」で安全に行う事が出来ます

yamadayy replied:
修正しました。

e-yamane replied:
ありがとうございます

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/simulation.py:
```diff
@@ -0,0 +1,70 @@
+    def __init__(self, t: list[Time] = [], w_list: list[WCSwId] = []):
+        self.__t = t
+        self.__w_list = w_list
+        self.__folder = 'tmp'
```
この手のオブジェクトプロパティをコンストラクタ内で固定かするとテストしにくかったりするので、コンストラクタ引数のデフォルト値とした方がいい気がしますね。特にここはフォルダ名ですし、、、

yamadayy replied:
変更しました。

e-yamane commented on src/telescope_baseline/tools/pipeline_v2/simulation.py:
```diff
@@ -0,0 +1,70 @@
+        sib = MapOnDetectorBuilder(9, pix_max, pix_max)
+        dib = DetectorImageBuilder(pix_max, pix_max, psf_w)
+        if not Path(self.__folder).exists():
+            Path(self.__folder).mkdir()
```
ここは、無視して頂いても構いませんが、本来的にはエラーハンドリングがいろいろ面倒なので、ディレクトリ生成は、parents=True（対称ディレクトリの親ディレクトリが無くてもたどって作ってくれる）とexist_ok=True(対称ディレクトリが既に存在してもエラーにはしない)は常に付けといても良いかもしれません。
この２つは普通にプログラム組んでて付けてて駄目なケースはほぼ思いつきません。

yamadayy replied:
直しました。

e-yamane replied:
ありがとうございます！！

e-yamane reviewed (commented):
第３弾コメント。

YoshiyukiYamada committed:
Move print sentence to the exception message.

YoshiyukiYamada committed:
Same sentences in diffrent method is defined in fixture.

YoshiyukiYamada committed:
Change from object loop to the loop of index because the index is used within the for loop.

YoshiyukiYamada committed:
Order of arguments in MapOnTheSky is changed. Id is the first argument.

YoshiyukiYamada committed:
Property self.__folder is fiven from constructor argument, not fixed value.

YoshiyukiYamada committed:
mkdir command is alwais do with parents=True and exists_ok=True options. Do not check existence etc.

e-yamane committed:
add saving test

YoshiyukiYamada committed:
Merge branch 'develop' into features/pipeline

YoshiyukiYamada committed:
Check only directory list, not file itself, for removing OS dependent error.

YoshiyukiYamada committed:
Codeclimate warning on PEP8 is removed. (complexity will be fixed in another commit)

YoshiyukiYamada committed:
remove unused attribute wcs from MapOnDetector. Add attribute wcs to MapOnTheSkyBuilder. Fix methods which calls these constructors.

YoshiyukiYamada committed:
For the specification consistency, static methods in builder are changed back to the class methods.

YoshiyukiYamada committed:
Apply cord climate indications.

YoshiyukiYamada committed:
Apply cord climate indications.

YoshiyukiYamada committed:
Apply cord climate indications.

yamadayy commented on tests2/test_detector_image.py:
```diff
@@ -0,0 +1,24 @@
+
+
+@pytest.fixture
+def di():
```
change name

yamadayy commented on tests2/test_map_on_the_sky_builder.py:
```diff
@@ -0,0 +1,52 @@
+
+
+@pytest.fixture
+def sc():
```
change name

yamadayy commented on tests2/test_position_on_detector.py:
```diff
@@ -0,0 +1,14 @@
+from telescope_baseline.tools.pipeline_v2.position_on_detector import PositionOnDetector
+
+
+def test_pos():
```
divide test method.

yamadayy commented on src/telescope_baseline/tools/pipeline_v2/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,131 @@
+        tupple[2].append(position_on_the_sky.coord.barycentricmeanecliptic.lat.rad)
+
+
+    @staticmethod
```
remove one blank line.

yamadayy commented on tests2/test_astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,38 @@
+
+import time
+
+def test_from_on_the_sky_position():
```
Add a blank line.

yamadayy commented on tests2/test_astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,37 @@
+    assert -3.9e-8 < coord[0][2] < -3.8e-8
+    assert -3.7e-10 < coord[0][3] < -3.6e-10
+    assert 4.8e-6 < coord[0][4] < 4.9e-6
+
```
remove it.

yamadayy commented on tests2/test_detector_image_catalogue.py:
```diff
@@ -0,0 +1,34 @@
+    assert sut_len == 7
+    with tempfile.TemporaryDirectory() as dname:
+        print(dname)
+        path = Path(dname, 'fits')  # 
```
remove it.

yamadayy commented on tests2/test_map_on_detector_builder.py:
```diff
@@ -0,0 +1,68 @@
+from telescope_baseline.tools.pipeline_v2.map_on_detector_builder import MapOnDetectorBuilder
+from telescope_baseline.tools.pipeline_v2.wcswid import WCSwId
+
+@pytest.fixture
```
Add a blank line

yamadayy commented on src/telescope_baseline/tools/pipeline_v2/map_on_detector_builder.py:
```diff
@@ -0,0 +1,74 @@
+            si.append(MapOnDetector(wl[0].wcs, positions_on_detector=a))
+        return si
+
+    def _store_list_of_detector_position(self, sky_positions, tmp):
```
extract method.

yamadayy commented on tests2/test_map_on_the_sky.py:
```diff
@@ -0,0 +1,54 @@
+    return m
+
+
+def test_save(mos):
```
divide method.

yamadayy commented on tests2/test_map_on_detector.py:
```diff
@@ -0,0 +1,60 @@
+
+@pytest.fixture
+def mod():
+    m = MapOnDetector([PositionOnDetector(1, Position2D(0.0, 0.0), Time('2000-01-01 00:00:00'), 12.5),])
```
Remove unnecessary comma.

yamadayy commented on tests2/test_map_on_the_sky.py:
```diff
@@ -0,0 +1,54 @@
+            assert r[4] == '2000-01-01 00:00:00.000'
+
+
+def test_load():
```
divide method.

yamadayy commented on tests2/test_position_on_the_sky.py:
```diff
@@ -0,0 +1,15 @@
+from telescope_baseline.tools.pipeline_v2.position_on_the_sky import PositionOnTheSky
+
+
+def test_pos():
```
divide method

yamadayy commented on tests2/test_astrometric_catalogue.py:
```diff
@@ -0,0 +1,45 @@
+from tests2.test_pipeline import get_tests_file_name
+
+
+def test_save():
```
divide method

yamadayy commented on tests2/test_astrometric_catalogue.py:
```diff
@@ -0,0 +1,45 @@
+    assert abs(float(r[7]) - 12.5) < 0.1
+
+
+def test_load():
```
divide method

yamadayy commented on tests2/test_detector_image_builder.py:
```diff
@@ -0,0 +1,33 @@
+from telescope_baseline.tools.pipeline_v2.map_on_detector import MapOnDetector
+
+
+def test_from_stellar_image():
```
divide method

yamadayy commented on tests2/test_astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,37 @@
+from telescope_baseline.tools.pipeline_v2.position_on_the_sky import PositionOnTheSky
+
+
+def test_from_on_the_sky_position():
```
divide method

yamadayy commented on tests2/test_map_on_detector.py:
```diff
@@ -0,0 +1,60 @@
+    assert m.datetime == '2000-01-01 00:00:00'
+
+
+def test_get_sky_positions(mod):
```
divide method

yamadayy commented on tests2/test_map_on_detector.py:
```diff
@@ -0,0 +1,60 @@
+    return m
+
+
+def test_save(mod):
```
divide method

yamadayy commented on tests2/test_map_on_detector.py:
```diff
@@ -0,0 +1,60 @@
+    assert p[4] == '2000-01-01 00:00:00.000'
+
+
+def test_load():
```
divide method

yamadayy commented on tests2/test_map_on_the_sky.py:
```diff
@@ -0,0 +1,63 @@
+    assert m.orbit_id == -1
+
+
+def test_load2(position_on_the_sky):
```
divide method

yamadayy commented on tests2/test_map_on_detector.py:
```diff
@@ -0,0 +1,71 @@
+    return next(iter(f))
+
+
+def test_save(csv_read):
```
divide method

yamadayy commented on tests2/test_map_on_the_sky.py:
```diff
@@ -0,0 +1,75 @@
+            assert int(r[0]) == -1
+
+
+def test_save2(file):
```
divide method

YoshiyukiYamada committed:
get_on_detector_posotions method in DetectorImage class is moved to MapOnDetectorBilder class.

YoshiyukiYamada committed:
Codeclimate PEP8 warning is resolved.

YoshiyukiYamada committed:
Move pipeline_v2 to pipeline. Also move tests2 to tests/pipeline.

YoshiyukiYamada committed:
Resolve PEP8 warning.

YoshiyukiYamada committed:
Resolve PEP8 warning.

YoshiyukiYamada committed:
Resolve PEP8 warning.

YoshiyukiYamada committed:
Fix assertion.

yamadayy commented:
レビューお願いいたします。

yamadayy commented:
Unresolved conversationのうち、山根さんコメントは山根さんがapproveしてくださったので私の方でresolve conversationとしておきました。一件残っているのは大澤さんのコメントですが、osではなくpathlibを使うようにというコメントで、対応済みです。これをresolveしていただければ、mergeできるようになると思います。

astronasutarou commented on src/telescope_baseline/tools/pipeline/astrometric_catalogue.py:
```diff
@@ -0,0 +1,51 @@
+        with open(file_name, 'w', newline='') as data_file:
+            write = csv.writer(data_file)
+            for e in self.__catalogue_entries:
+                write.writerow([e.stellar_id, e.coord.icrs.ra.deg, e.coord.icrs.dec.deg,
+                                e.coord.icrs.pm_ra_cosdec * u.yr / u.mas, e.coord.icrs.pm_dec * u.yr / u.mas,
+                                e.coord.distance / u.pc, e.coord.obstime, e.mag])
```
`* u.yr / u.mas` 等は無次元化しているという理解であっていますか？こうしたケースでは `to_value()` method を使うほうが安全かと思います．

- https://docs.astropy.org/en/stable/units/quantity.html

astronasutarou replied:
ついでにちょっと確認です．`AstrometricCatalog` には観測時刻が必須なのでしたっけ？ある epoch の (ra, dec, v_ra_cosdec, v_dec, parallax) が記録されており，観測時刻 (と軌道情報) が定義されて `PositionOnTheSky` に変換されるのだと思って資料を読んでいたので少し引っかかりました．

yamadayy replied:
to_value()知りませんでした。修正します。AstrometricCatlogueに時刻が入っているのは、時刻が無いとpositionがどの時点のposition だかわからないからです。proper motionで動くので。ですので、PositionOnTheSkyと同じ意味の時刻ではありませんが、時刻が必要です。良いネーミングがあれば提案してください。

astronasutarou replied:
> 時刻が無いとpositionがどの時点のposition だかわからないからです。proper motionで動くので。

状況を理解しました．カタログでは epoch が共通であるということを想定していたので，個別の天体に対して時刻を設定していることに引っかかったのでした．csv として保存している現状ではこの実装でいいと思います．

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image_builder.py:
```diff
@@ -0,0 +1,51 @@
+        self.__ny = ny
+        self.__psf_w = psf_w
+
+    def from_stellar_image(self, si: MapOnDetector) -> DetectorImage:
+        """Build DetectorImage from StellarImage class
```
名称と説明が一致していないような気がするのでご確認をお願いします．

yamadayy replied:
修正します。

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image_builder.py:
```diff
@@ -0,0 +1,51 @@
+        return di
+
+    def _generate_a_stellar_image(self, a, dps):
+        for k in range(int(dps.mag)):
```
これは `mag` が photon 数を意味しているのでしたっけ？

magnitude の意味を考えると misleading なので `CatalogueEntry` に magnitude → photon 数を計算する method を実装しておいたほうがいいと思います．

astronasutarou replied:
この PR で修正は求めませんがこの実装だと以下のような問題があると思います．

- photon の Poisson ノイズが考慮されていない
- 明るい星に対して極端にパフォーマンスが落ちる

yamadayy replied:
済みません。現実装ではmagに光子数を入れてしまっています。この変換を考えるのは、いろんなパラメータに依存して面倒なので、後回しにさせてください。とりあえず、magを返すだけのget_nphotonを作ってかぶせておいて、TODOをつけました。
明るい星のパフォーマンスはご指摘の通りです。xp, ypの計算に乱数を使っているので、Poiossonノイズは自然に入ると考えています。

astronasutarou replied:
> Poiossonノイズは自然に入ると考えています。

もし勘違いをしていたら申しわけないのですが，乱数によって位置が変わるので pixel 単位でのカウントはばらつきますが総 photon 数が保存しているので (特に暗い側で) 統計がおかしくなるような気がします．

yamadayy replied:
私の理解は、乱数が確率分布で与えられている、したがって、あるセルに光子が入る確率pはこの分布で表されている。そこにN個の光子がばらまかれる、したがって、実際にセルに光子が入るか入らないかはPoisson過程で、これで数えられる光子数はPoisson誤差を入れたものになっている、です。確率分布は無限に遠いところまで見ているので、総photon数が一定であることは悪いことをしないと思うのですが(星像Windowに入る数もPoisson誤差が入っているはず)、何か勘違いしていますでしょうか？

astronasutarou replied:
> 確率分布は無限に遠いところまで見ているので、総photon数が一定であることは悪いことをしないと思うのですが

とりあえず測光したときの photon noise がほぼゼロになるので現実を反映していないような……．

N が大きいときに各ピクセルのカウントは近似的に Poissson 過程に従うとは思うのですが，ピクセル間に相関が生まれてしまっているので和をとった後の結果が統計的に歪むのだと思います．上で挙げた測光もそうですが中心位置推定にも影響がありそうに思いました．

yamadayy replied:
論点理解しました。空間分布としてPoissonノイズが十分入っているはずだから大丈夫だと思っていましたが、時間軸側でも光子総数n個に対するPoisson分布があるから、総数が固定ではいけないというご指摘ですね。そんな気がしてきました。総数nはもっと大きな数Nに対する確率pがあって、n=Npで決まるはずなのですが、Nが何なのか、pが何なのかがしっくり来ていませんが、とりあえず総数を√Nで揺らがせて与えることにします。

astronasutarou replied:
`numpy.random.poisson` を使ってください．

yamadayy replied:
修正しました。

astronasutarou replied:
修正を確認しました．

Poisson distribution の range は正整数なので負値チェックはなくても大丈夫です．手元で調べたら返り値の型も整数だったのでキャストも不要かもしれません．

とりいそぎ close します．

yamadayy replied:
修正しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+from astropy.wcs import WCS
+
+
+class MapOnDetectorBuilder:
+    """Builder class for Stellarimage class
```
説明が過去の名称を引きずっているでしょうか．

yamadayy replied:
修正します。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+
+        Args:
+            wcs: world coordinate system
+            c: DtectorImageCatalogue
```
s/DtectorImageCatalogue/DetectorImageCatalogue/

yamadayy replied:
修正します。

astronasutarou commented on tests/pipeline/data/a.csv:
```diff
@@ -0,0 +1 @@
+1,266.4049882865447,-28.936177761791473,-3.325090692327255,13.745681936077188,1.0,2000-01-01 00:00:00.000,12.5
```
テスト用のファイルではありますがファイル名を再考してほしいです．

astronasutarou replied:
冗長かもしれませんがカラム名があったほうがいいと思いました．

astronasutarou replied:
カラム名については `sky.csv` との consistency を考えて不要だと思い直しました．

yamadayy replied:
ファイル名を変更しました。

astronasutarou commented on tests/pipeline/test_pipeline.py:
```diff
@@ -0,0 +1,71 @@
+    w.wcs.ctype = ["GLON-TAN", "GLAT-TAN"]
+    wlist = []
+    for i in range(len(t)):
+        wlist.append(WCSwId(i, 1, copy.deepcopy(w)))
+    c = Simulation(t=t, w_list=wlist)
+    b = c.do(a, 1024, 1.0)
+    assert(b is not None)
```
異常時に `None` が返ってくる実装になっていないと思います．

yamadayy replied:
今後の課題として、テストを追加します。とりあえず、これはオブジェクトを作ったらOK、作ったオブジェクトが異常でもOKというテストです。

astronasutarou replied:
……すみません．僕のコメントがまったく指示になっていませんでした．

`assert(b is not None)` と書くと `c.do()` の返り値が `None` ではないことが正常だという表明になります．現在の実装ではそうなっていません．混乱を生むので assert 文は削除してください．

yamadayy replied:
修正しました。

astronasutarou commented on tests/pipeline/test_pipeline.py:
```diff
@@ -0,0 +1,71 @@
+    t = hdu.header['DATE-OBS']
+    w = astropy.wcs.WCS(hdu.header)
+    a = c.do(d, w)
+    assert(a is not None)
```
異常時に `None` が返ってくる実装になっていないように思います．

yamadayy replied:
simulationのテストと同様、簡易テストです。あとで考えます。

astronasutarou replied:
……すみません．僕のコメントがまったく指示になっていませんでした．

`assert(b is not None)` と書くと `c.do()` の返り値が `None` ではないことが正常だという表明になります．現在の実装ではそうなっていません．混乱を生むので assert 文は削除してください．

yamadayy replied:
修正しました

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+            sil.append(MapOnDetector(self.get_on_detector_positions(di, self.__window_size)))
+        return sil
+
+    def from_on_tye_sky_position(self, o: MapOnTheSky, wl: list[WCSwId]) -> list[MapOnDetector]:
```
s/from_on_tye_sky_position/from_on_the_sky_position/

yamadayy replied:
修正しました

astronasutarou commented on tests/pipeline/data/detector.csv:
```diff
@@ -0,0 +1 @@
+1,0.0,0.0,12.5,2000-01-01 00:00:00
```
カラム名 (ヘッダー行) を加えませんか？

astronasutarou replied:
あー…… `sky.csv` では 1 行目だけ例外 (orbit id) として扱う実装をしているのですね．
consistency がとれないのでヘッダは要らないと思い直しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/wcswid.py:
```diff
@@ -0,0 +1,31 @@
+from astropy.wcs import WCS
+
+
+class WCSwId:
```
orbit & exposure と望遠鏡の指向方向を結びつける object という理解でいいでしょうか？
ちょっと名前と機能に隔たりがあるように思いました (修正依頼ではないです)．

yamadayy replied:
考えておきます。

astronasutarou commented on tests/pipeline/test_detector_image.py:
```diff
@@ -0,0 +1,16 @@
+
+
+def test_load(detector_image):
+    assert detector_image is not None
```
失敗したときに `load` 関数は `None` を返さないように思います．

yamadayy replied:
修正しました。

astronasutarou commented on tests/pipeline/test_pipeline.py:
```diff
@@ -0,0 +1,71 @@
+
+
+def get_tests_file_name(fname: str, folder='data'):
+    a = Path.cwd()
+    tests_dir = 'tests/pipeline'    # change it when test dir move
```
`Path.cwd()` で返ってくる値としてはどのようなケースを想定しているのでしょうか．

ある決まったディレクトリから実行しないと正しいファイルパスが取得できない実装になっているでしょうか？実際にはルート以外で実行するケースはほとんどないと思いますが少し気になりました．

'tests/pipeline' 以下で作業をするのでしたら，このファイル (`test_pipeline.py`) の場所を取得して相対パスを作成するという手続きにしたほうが安心かもしれないと思いました．

yamadayy replied:
pycharmで実行するとtests/pipelineで、github actionで実行するとsrcで、手動で実行する場合は普通はトップディレクトリーになると思います。その違いを吸収しようとしています。

e-yamane replied:
ここは、私も気になってましたが、pipelineが長期生存していて、actionsの進化に追いついてないんですよね。
ここは、developにマージ後に私の方でシンプルにしようと思ってました。
＃コメント足しとけば良かったなぁ...

ちなみに、
> pycharmで実行するとtests/pipeline
これ本当ですか？
IntelliJとかでJavaのテスト実行するときってカレントディレクトリはプロジェクトルートなイメージありますが、、、

astronasutarou replied:
ちなみに僕のローカル環境ではルートディレクトリの名前が `telescope_baseline` ではなく `baseline` になっているために正常に動きませんでした．想定外のケースですね．

まあ GitHub actions で動いているのでこのコメントは close します．

yamadayy replied:
プロジェクトルートをとる方法もあるみたいなので、次のチケットで考えてみます。

astronasutarou commented on tests/pipeline/test_pipeline.py:
```diff
@@ -0,0 +1,71 @@
+        return Path(a, tests_dir, folder, fname)
+    while a.name != 'telescope_baseline':
+        a = a.parent
+    if folder != 'data':
+        tmp_path = Path(a, tests_dir, folder)
+        if not tmp_path.exists():
+            tmp_path.mkdir()
```
ファイル名を取得する関数が (temporary とはいえ) ディレクトリを作成する機能を持つことには違和感があります．一時ファイル or 一時ディレクトリを作成する機能は別の関数として実装しませんか．

astronasutarou replied:
一時ファイル作成機能は消すところまで実装したいですがテストと連携させる方法は調査の必要がありそうですね．

yamadayy replied:
調査後、ということでお願いします。

astronasutarou replied:
了解しました．

astronasutarou commented on tests/pipeline/test_map_on_detector.py:
```diff
@@ -0,0 +1,79 @@
+
+@pytest.fixture
+def csv_read(mod):
+    f_name = get_tests_file_name('a.csv', folder="tmp")
+    mod.save(f_name)
+    file = open(f_name, 'r', newline='')
+    f = csv.reader(file, delimiter=',')
+    return next(iter(f))
```
ファイルシステムまで含めたテストが必要ということでなければ IO stream を使用すると一時ファイルを作成せずに済むかもしれません．

- https://docs.python.org/3/library/io.html

yamadayy replied:
次の課題としてください。

astronasutarou replied:
了解しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+
+        Returns: list of positions in detector coordinate.
+
+        """
```
PSF を用いた測定をしているので get よりも重たい単語を使いたいと思いました．

説明の Calculate image center position は何を測っているのかが不明瞭なので書き直したほうが良さそうです．

astronasutarou replied:
大きい画像から位置測定までがひとつの関数で実行されていますが，機上での操作と地上での操作が両方含まれているので分割したいです．現在の設計に加えて 9×9 の切り抜き画像というデータクラスがあったほうがいいと思います．

yamadayy replied:
切り抜きクラスの作成については次の課題とさせてください。説明の書き直しはしました。getでないとすると、estimateくらいいでしょうか？

astronasutarou replied:
感覚に依るところが大きそうですがぱっと思いついたのは estimate とか measure とかでしょうか……．

説明の書き換えを確認しました．個人的な意見ですが `stellar window` と `image center position` が何を指しているのか分かりにくいので，他にコメントを見直す機会があれば再考をお願いします．

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+        e_psf_model, e_psf_stars = e_psf_builder(es)
+        # TODO need to implement cross-match
+        position_list = []
+        for s in e_psf_stars.all_stars:
+            # TODO IDもこれじゃ駄目〜
+            # TODO 明るさの計算はまだ（0にしてる）
+            # TODO 時刻をHBUから取ってる来る処理がまだ！！
+            position_list.append(
+                PositionOnDetector(1, Position2D(s.center[0], s.center[1]), Time('2000-01-01 00:00:00.0'), mag=3000))
```
exposure ID, 時刻は画像データ (DetectorImage) に含まれている (あとはとってくるだけ) という想定だと認識しましたが正しいでしょうか．

yamadayy replied:
はい。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+        e_psf_builder = EPSFBuilder(oversampling=4, maxiters=3, progress_bar=True)
+        es = e_psf_stars
+        e_psf_model, e_psf_stars = e_psf_builder(es)
+        # TODO need to implement cross-match
```
PositionOnDetector には天体を特定するための情報が含まれていない設計になっていると思いますが，この段階で天体の同定 (cross match) が必要になるという想定でしょうか？

yamadayy replied:
どこにクロスマッチを入れるかはもう少し考えさえてください。

astronasutarou replied:
了解しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/position_on_detector.py:
```diff
@@ -0,0 +1,42 @@
+    """Data class for detector position of individual image.
+
+    """
+    def __init__(self, exposure_id: int, p: Position2D, datetime: Time, mag: float = 12.5):
```
`mag` のデフォルト値は必須でしょうか？
設定する場合も「有り得そうな値」は避けておいたほうが無難かもしれません．

yamadayy replied:
消しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/position_on_detector.py:
```diff
@@ -0,0 +1,42 @@
+
+
+class PositionOnDetector:
+    """Data class for detector position of individual image.
```
個々の画像 (検出器) における天体の位置を保存するクラスではないでしたっけ？
この説明ではちょっと誤解を生むような気がします．

yamadayy replied:
修正してみました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image.py:
```diff
@@ -0,0 +1,34 @@
+        return DetectorImage(ft[0])
+
+    def save(self, filename: str) -> None:
+        self.__hdu.writeto(filename, overwrite=True)
```
`overwrite` は `True` をベタ書きせずオプションで設定できるようにしてください．

yamadayy replied:
修正しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image_builder.py:
```diff
@@ -0,0 +1,51 @@
+        di = DetectorImage(hdu)
+        return di
+
+    def _generate_a_stellar_image(self, a, dps):
```
"星像を作る関数" は個々の星に対応するオブジェクトに持たせたほうがいいような気がします．

yamadayy replied:
そのようなオブジェクトが今は無いので、次の課題とさせてください。

astronasutarou replied:
了解しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/astrometric_catalogue.py:
```diff
@@ -0,0 +1,51 @@
+    Attributes:
+        catalogue_entries: list of CatalogueEntry which is 5 parameter list for one star.
+
+    """
```
等級があるのでパラメタの数は 5 ではないような……

yamadayy replied:
コメント修正しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/simulation.py:
```diff
@@ -0,0 +1,67 @@
+        Returns:
+            A DetectorImageCatalogue object which contains DetectorImage objects of whole mission.
+
+        """
```
別の場所でもコメントしましたが等級を含んでいるので `AstrometricCatalogue` のパラメタ数は 6 ではないでしょうか．また `pixel number` よりは画像サイズ (検出器サイズ？) と説明してほしいです．

return するのは `DetectorImageCatalogue` ではなくてその List ですね．

yamadayy replied:
修正しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/simulation.py:
```diff
@@ -0,0 +1,67 @@
+
+    """
+
+    def __init__(self, t: list[Time] = [], w_list: list[WCSwId] = [], folder: str = 'tmp'):
```
ここのデフォルト値って必要でしょうか？後から時刻やポインティングリストを編集することはなさそうな気がします．

yamadayy replied:
フォルダー名以外はdefault値を消しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,101 @@
+    for i in range(len(ty)):
+        lont = lon0 + (para * math.sin(ls[i] - lon0) + pm_lon_coslat * ty[i]) / math.cos(lat0)
+        latt = lat0 + (pm_lat * ty[i] - para * math.sin(lat0) * math.cos(ls[i] - lon0))
+        residual[i] = (lon[i] - lont) ** 2 + (lat[i] - latt) ** 2
```
ここはベクトル演算で書いたほうがいいと思います．

yamadayy replied:
スカラーが絡んでいるので、どうやるのかすぐに調べられませんでした。(できませんと書いている記事もある)。保留させてください。

astronasutarou replied:
了解しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,101 @@
+    ty = []
+    for time in t:
+        ls.append(get_sun(time).geocentricmeanecliptic.lon.rad)
+        ty.append(time.jyear)
```
また `get_sun(time)` も時刻のアレイを引数にとれるはずなので for-loop 回さなくてもできるはずです．また，`astropy.time.Time` は複数の時刻をアレイとして持つことができたと思います．その実装を使うとユリウス日等への変換も loop を回さずに済みます．

yamadayy replied:
少し格闘したのですが、うまくいきませんでした。あとで調べますが、取りあえずこのままでお願いします。

astronasutarou replied:
今後検討ということで了解しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,101 @@
+        tupple[2].append(position_on_the_sky.coord.barycentricmeanecliptic.lat.rad)
+
+    @staticmethod
+    def from_on_the_sky_position(otsp: list[MapOnTheSky]):
+        dic = {}
+        for map_on_the_sky in otsp:
+            for position_on_the_sky in map_on_the_sky.positions_on_the_sky:
+                AstrometricCatalogueBuilder._update_dic(dic, position_on_the_sky)
+        result = []
+        for t, londata, latdata in dic.values():
+            # TODO. Input appropriate initial value for least square.
+            parameter = [np.radians(266), np.radians(-5), 0., 0., np.radians(1 / 3600)]
+            result.append(optimize.leastsq(lsf_fit_function_for_astrometric_parameters, parameter,
+                                           args=(t, londata, latdata)))
+        return AstrometricCatalogue(result)
```
`staticmethod` で `A.from_...` というメソッドを実行した場合には返り値が `A` であることを期待してしまう気がします．今回は `AstrometricCatalogBuilder.from_on_the_sky_position` という名前のメソッドが `AstrometricCatalog` を返すので引っかかりを感じました．

yamadayy replied:
xxBuilderはxxを作るためのクラスという位置づけですので、ABuilderがAを返すのは設計通りです。ABuilderを返してはいけません。(ということでご指摘の答えになっているでしょうか？)

astronasutarou replied:
すみません！ `@classmethod` と勘違いをして的はずれなコメントをしていました．

個人的には `build` や `generate` みたいな動詞が欲しくなりましたが Builder クラスが build するのは当然なので要らないのですかね．resolved にします．

astronasutarou commented on src/telescope_baseline/tools/pipeline/analysis.py:
```diff
@@ -0,0 +1,51 @@
+            else:
+                dic_list.append(DetectorImageCatalogue(di))
+                di = [cat[i]]
+        dic_list.append(DetectorImageCatalogue(di))
```
見慣れない条件が出てきて戸惑ったのですが，これは軌道周回ごとにデータを整理しているという認識で正しいでしょうか？実際には各画像に軌道 ID が振られておりそれで分類するという実装になるのですかねぇ……．

yamadayy replied:
はい、そうです。

astronasutarou commented on src/telescope_baseline/tools/pipeline/analysis.py:
```diff
@@ -0,0 +1,51 @@
+
+        Returns:
+            AstrometricCatalogue which contains list of 5 parameters of whole stars.
+        """
```
現時点では仕方ないのだと思いますが `wcs` (ポインティング情報) が `Analysis` の入力にあるのはちょっと引っかかりました．現在の構成なら `MapOnDetectorBuilder` が生成する役割を担うと思います．`window_size` も本来であればダウンリンクされたデータから自然に決まる値なので引数にはならないと思います (解析としては運用中に変わってもいい)．

データ設計次第ですが望遠鏡の指向方向 (暫定値) などのメタ情報は fits ファイルのヘッダから読み取ることになるのかなと思っています．

yamadayy replied:
はい、ご指摘の通りです。TODOで書いておきます。

astronasutarou commented on src/telescope_baseline/tools/pipeline/simulation.py:
```diff
@@ -0,0 +1,67 @@
+            di[j].hdu.header['DATE-OBS'] = str(self.__t[o.orbit_id])
+            fname = "tmp" + str(o.orbit_id) + "_" + str(j) + ".fits"
+            fpathname = Path(self.__folder, fname)
+            di[j].hdu.writeto(str(fpathname), overwrite=True)
```
経験的に `overwrite` のデフォルトを `True` にするのはやめたほうがいいと思います．

yamadayy replied:
修正しました。

astronasutarou replied:
修正を確認しました．

astronasutarou commented on tests/pipeline/test_astrometric_catalogue_builder.py:
```diff
@@ -0,0 +1,34 @@
+from telescope_baseline.tools.pipeline.position_on_the_sky import PositionOnTheSky
+
+
+def test_from_on_the_sky_position1():
+    t = [Time('2000-01-01 00:00:00.0'), Time('2000-02-01 00:00:00.0'), Time('2000-03-01 00:00:00.0'),
+         Time('2000-04-01 00:00:00.0'), Time('2000-05-01 00:00:00.0'), Time('2000-06-01 00:00:00.0'),
+         Time('2000-07-01 00:00:00.0'), Time('2000-08-01 00:00:00.0'), Time('2000-09-01 00:00:00.0'),
+         Time('2000-10-01 00:00:00.0'), Time('2000-11-01 00:00:00.0'), Time('2000-12-01 00:00:00.0'),
+         Time('2001-01-01 00:00:00.0')]
+    londata = [4.657229292609037, 4.657231613943884, 4.6572328740589235, 4.657232905752553, 4.657231693733922,
+               4.657229540136522, 4.657227130343018, 4.657224922535985, 4.65722356203127, 4.6572234180551595,
+               4.657224576568599, 4.657226710038408, 4.657229355524401]
+    latdata = [-0.09662669750312813, -0.09662682006310001, -0.0966270232593704, -0.09662727212328368,
+               -0.09662747861779832, -0.09662760274710183, -0.0966276096289789, -0.09662749965491312,
+               -0.09662729760688273, -0.09662706177588641, -0.09662684008417984, -0.09662670773667688,
+               -0.09662669894287566]
+    builder = AstrometricCatalogueBuilder()
+    o = []
+    for i in range(len(t)):
+        s = PositionOnTheSky(1, SkyCoord(lon=londata[i], lat=latdata[i], unit=('rad', 'rad'),
+                                         frame='barycentricmeanecliptic'), 3000, t[i])
+        o.append(MapOnTheSky(positions_on_the_sky=[s]))
+    a2 = builder.from_on_the_sky_position(o)
```
この関数の挙動をテストするためには星を複数個含むデータを使用するべきかなと思いました．

yamadayy replied:
今後の課題としてください。

astronasutarou replied:
せめてもうひとつ用意しませんか？修正範囲はそれほど多くないと思うのですが……．

`AstrometricCatalogueBuilder` のコアとなる機能に星を stellar ID で分類する関数がありますが，星が複数あることではじめてその機能のテストになると思うので……．

yamadayy replied:
上でのIDの代入のところで暫定で1を入れていてIDの付け方を今後の課題にしています。ですので、2つでもここを解決しないといけないので、これは今後の課題とさせてもらえないでしょうか。

astronasutarou replied:
`AstrometricCatalogBuilder` の単体テストであれば ID の問題には引っかからないと思ったのですが，何か見落としがあるでしょうか？試しに手元で同じデータを Stellar ID = 2 として入れてみましたが想定した挙動をしていました．

参考までに patch を書いておきます．

``` patch
--- a/tests/pipeline/test_astrometric_catalogue_builder.py
+++ b/tests/pipeline/test_astrometric_catalogue_builder.py
@@ -26,10 +26,19 @@ def test_from_on_the_sky_position1():
         s = PositionOnTheSky(1, SkyCoord(lon=londata[i], lat=latdata[i], unit=('rad', 'rad'),
                                          frame='barycentricmeanecliptic'), 3000, t[i])
         o.append(MapOnTheSky(positions_on_the_sky=[s]))
+    for i in range(len(t)):
+        s = PositionOnTheSky(2, SkyCoord(lon=londata[i], lat=latdata[i], unit=('rad', 'rad'),
+                                         frame='barycentricmeanecliptic'), 3000, t[i])
+        o.append(MapOnTheSky(positions_on_the_sky=[s]))
     a2 = builder.from_on_the_sky_position(o)
 
-    coord = a2.get_catalogue()[0]
+    coord = a2.get_catalogue()
+
     # TODO: need to fix proper motion check. Windows and Linux give different answer.
-    assert 4.657 < coord[0][0] < 4.658
-    assert -0.1 < coord[0][1] < -0.09
-    assert 4.8e-6 < coord[0][4] < 4.9e-6
+    assert 4.657 < coord[0][0][0] < 4.658
+    assert -0.1 < coord[0][0][1] < -0.09
+    assert 4.8e-6 < coord[0][0][4] < 4.9e-6
+
+    assert 4.657 < coord[1][0][0] < 4.658
+    assert -0.1 < coord[1][0][1] < -0.09
+    assert 4.8e-6 < coord[1][0][4] < 4.9e-6
```

yamadayy replied:
MapOnTheSkyのID付け問題と関係ないということで了解しました。修正します。

astronasutarou replied:
テストのアップデートを確認しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/wcswid.py:
```diff
@@ -0,0 +1,31 @@
+            orbit_id: orbit ID
+            exposure_id: exposuer ID
+            wcs: World Coordinate System
+        """
```
`orbit_id` は周回ごとに変わる通し番号で，`exposure_id` は周回内でカウントアップする通し番号という理解でよいでしょうか？

astronasutarou replied:
他のところを見ているとそうではないような感じがしますね……撮像のたびに (衛星の位置が変わるたびに) `orbit_id` も変わるという認識で正しいでしょうか？

yamadayy replied:
orbit_idは周回ごと、exposure_idは撮像ごとの通し番号です。exposure_idは周回内だけでなく、周回をまたいでどんどんカウウントアップすることを想定しています。テストでは、一周回で一撮像しかしていないので、分かりにくいですね。

astronasutarou commented on src/telescope_baseline/tools/pipeline/simulation.py:
```diff
@@ -0,0 +1,67 @@
+                wl.append(w)
+        return wl
+
+    def _generate_one_hdu(self, dib, o, sib, wl):
```
`_generate_one_hdu` という関数が `len(wl)` 個のファイルを生成するのはあまり良い naming ではないと思いました．

yamadayy replied:
修正しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_the_sky_builder.py:
```diff
@@ -0,0 +1,79 @@
+
+        Returns:
+
+        """
```
リファクタリング中に修正しそこねた関数名でしょうか？

yamadayy replied:
修正しました。

astronasutarou replied:
`from_astrometric_catalogue_2_list` の 2 は必要でしょうか．

yamadayy replied:
外します。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_the_sky_builder.py:
```diff
@@ -0,0 +1,79 @@
+        for position in mod.positions_on_detector:
+            sky = self.__wcs.pixel_to_world(position.x, position.y)
+            # TODO: Consideration is needed how ids are set.
+            ret.append(PositionOnTheSky(position.exposure_id, sky, position.mag, position.datetime))
```
**To Do** であることは理解しますが，暫定的な処置であっても "それっぽいもの" が入っていると混乱するので stellar ID であるところに exposure ID を入れるのはやめてください．

yamadayy replied:
修正しました

astronasutarou replied:
確認させてください．更新されたコードでは `stellar_id = position.exposure_id` として名前を付け替えていますが，現在のコードでは `position.exposure_id` を代入しないと動かない状態になっているのでしょうか．仮に何か問題が発生したときに情報の伝達が切れていたほうが良いと思うのですが，固定値ではダメなのでしたっけ？

暫定的であったとしても意味があるコードを書かれると何らかの表明に見えてしまいます．今回のケースではさほど問題にならないと思いますが「ID を ID に渡している」という状況は一見まともに見えかねないのですごく嫌な臭いがするのです．

yamadayy replied:
暫定で1を入れます。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_the_sky_builder.py:
```diff
@@ -0,0 +1,79 @@
+    def __init__(self, wcs: WCS):
+        self.__wcs = wcs
+
+    def get_sky_positions(self, mod: MapOnDetector):
```
実際には同じ周回内の `List[MapOnDetector]` をまとめて処理をすることになると思います．軌道周回ごとのデータをまとめるクラスがあったほうが良さそうに思いました．

yamadayy replied:
クラスの定義は、どういうまとまりかわからないけどまとまりということだけで、どうまとめるかはまとめる側の問題として処理する、ではだめでしょうか。

astronasutarou replied:
"まとまり" に機能をもたせたくなるかどうか……ですかね．
なんとなく必要になるかなと思いましたが実際にやってみないとわからないですね．

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_the_sky_builder.py:
```diff
@@ -0,0 +1,79 @@
+    # TODO: unit of distance and proper motion is assumed to be 'pc' and 'mas/yr' respectively.  It should be checked.
+    para = 1 / coord.distance.value * 4.848136e-6  # const is conversion ratio from as to rad.
+    pm_lon_coslat = coord.barycentricmeanecliptic.pm_lon_coslat.value * 4.848136e-9
+    pm_lat = coord.barycentricmeanecliptic.pm_lat.value * 4.848136e-9
```
`astropy.Quantity` を使用しているのであれば単位変換は `to.("target unit")` でできると思います．

yamadayy replied:
ありがとうございます。修正しました。

astronasutarou replied:
確認しました．
// 数値が欲しいだけなので `to_value("target unit")` を使うようコメントした方が良かったですね……

astronasutarou commented on tests/pipeline/test_pipeline.py:
```diff
@@ -0,0 +1,71 @@
+def test_analysis():
+    c = Analysis()
+    d = DetectorImageCatalogue([
+        DetectorImage.load(str(get_tests_file_name('tmp1_0.fits'))),
+        DetectorImage.load(str(get_tests_file_name('tmp2_0.fits'))),
+        DetectorImage.load(str(get_tests_file_name('tmp7_0.fits'))),
+        DetectorImage.load(str(get_tests_file_name('tmp3_0.fits'))),
+        DetectorImage.load(str(get_tests_file_name('tmp8_0.fits'))),
+        DetectorImage.load(str(get_tests_file_name('tmp9_0.fits')))
```
テスト用のデータも役割がわかるように命名してください．

yamadayy replied:
この一連の指摘は、後日対応とさせてください。いくつかは勝手にresolveしています。

astronasutarou replied:
了解しました．

astronasutarou commented on tests/pipeline/test_map_on_detector_builder.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def test_from_detector_image_catalogue():
+    f_name = str(get_tests_file_name('tmp.fits'))
```
テスト用のデータも機能がわかるように命名してください．

astronasutarou commented on tests/pipeline/test_detector_image.py:
```diff
@@ -0,0 +1,16 @@
+
+@pytest.fixture
+def detector_image():
+    return DetectorImage.load(get_tests_file_name('tmp0_0.fits'))
```
テスト用のデータも役割がわかるように命名してください．

astronasutarou commented on tests/pipeline/test_map_on_detector_builder.py:
```diff
@@ -0,0 +1,79 @@
+
+def test_position():
+    builder = MapOnDetectorBuilder(9, 256, 256)
+    detector_image = DetectorImage.load(get_tests_file_name('tmp0_0.fits'))
```
テスト用のデータも役割がわかるように命名してください．

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+
+        """
+        data = detector_image.hdu.data
+        peaks_tbl = find_peaks(data, threshold=200.)
```
`threshold` はベタ書きせず引数で与えるようにしてください．

yamadayy replied:
パラメータ数が4を超えるので、TODOにしました。いくつかのパラメータをクラス(C++なら構造体でも良いのだけど)にまとめる必要があり、設計変更になります。

astronasutarou replied:
TODO ということで了解しました．

これをクラスにまとめて引数にするかと言われると微妙ですね…… Builder の attribute として与えたほうがスッキリする気がします．Code Climate の引数の制限は結構厳しいので調整するか，一部の関数では wontfix 宣言をして引数の超過を許容しても良いんじゃないかなと思います．

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+        """
+        data = detector_image.hdu.data
+        peaks_tbl = find_peaks(data, threshold=200.)
+        half_size = (window_size - 1) / 2
```
`window_size` が奇整数であることが前提になっていたりしますか？

yamadayy replied:
はい、前提にしています。

astronasutarou replied:
重箱の隅をつつくようですがどこかに assertion を書いておきませんか？

yamadayy replied:
assertを置きました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,118 @@
+        nddata = NDData(data=data)
+        e_psf_stars = extract_stars(nddata, stars_tbl, size=window_size)
+        # TODO need to convert A/D value to photon count
+        e_psf_builder = EPSFBuilder(oversampling=4, maxiters=3, progress_bar=True)
```
このあたりのパラメタも関数の引数として設定できるようにしてください．

yamadayy replied:
修正しました。(commitすると、引数の数が多すぎるエラーが出ると思いますが)

yamadayy replied:
Code climateがエラーを出すので、progress_barは引数から外しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/catalog_entry.py:
```diff
@@ -0,0 +1,58 @@
+
+    @property
+    def distance(self):
+        return self.__coord.distance / u.pc
```
単位なしの数値にしたい場合は `.to_value("target unit")` メソッドを使うのが安全です (以下同様)．

yamadayy replied:
修正しました。

astronasutarou replied:
`distance`, `pm_ra_cosdec`, `pm_dec` の修正もお願いします．

yamadayy replied:
済みません。確認忘れです。修正しました。

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image_builder.py:
```diff
@@ -0,0 +1,51 @@
+            ny: array size of y direction
+            psf_w: psf width
+        """
+        self.__nx = nx
+        self.__ny = ny
```
nx = NAXIS1, ny = NAXIS2 という理解でいいでしょうか？
その場合，現在の実装では x, y が逆になる箇所がいくつかあります．

astronasutarou replied:
`nx`, `ny` よりは FITS の定義に寄せて  NAXIS1, NAXIS2 に近い名称 (例えば `naxis1`, `naxis2`) にしたほうが誤解がないかもしれません．

yamadayy replied:
アレイサイズをarray[nx][ny]としていて、FITSにこれを渡すと、naxis1=ny, naxis2 = nxとなると思います。アレイの扱いを基準にしたつもりですが、問題があれば、今後設計議論をさせてください。取りあえずこのままで。

astronasutarou replied:
とりあえずこのままということで了解しました．

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image_builder.py:
```diff
@@ -0,0 +1,51 @@
+
+        """
+        dp = si.positions_on_detector
+        a = np.random.uniform(0.0, 10.0, (self.__nx, self.__ny))
```
`nx`, `ny` の定義にも依りますが `(self.__ny, self.__nx)` が正しいと思います．

astronasutarou replied:
手元で実行した結果です．

``` python
import numpy as np
import astropy.io.fits as fits

data = np.zeros((10, 50))
hdu = fits.PrimaryHDU(data=data)
hdu.data.shape
--->
  (10, 50)
hdu.header
--->
  SIMPLE  =                    T / conforms to FITS standard                      
  BITPIX  =                  -64 / array data type                                
  NAXIS   =                    2 / number of array dimensions                     
  NAXIS1  =                   50                                                  
  NAXIS2  =                   10                                                  
  EXTEND  =                    T                                                  
```

yamadayy replied:
はい、逆になることは認識しています。これを変更しようとすると全体を見ないといけないので、設計議論後でお願いします。

astronasutarou commented on src/telescope_baseline/tools/pipeline/detector_image_builder.py:
```diff
@@ -0,0 +1,51 @@
+
+    def _generate_a_stellar_image(self, a, dps):
+        for k in range(int(dps.mag)):
+            xp = int(self.__psf_w * np.random.randn() + dps.y + 0.5)
+            yp = int(self.__psf_w * np.random.randn() + dps.x + 0.5)
+            if 0 <= xp < self.__nx and 0 <= yp < self.__ny:
+                a[xp][yp] += 1
```
FITS 画像の (NAXIS1, NAXIS2) = (nax1, nax2) にアクセスするためには `data[nax2][nax1]` となります．x, y を flip させると混乱するので，個人的には画像では `x`, `y` を使用せず以下のように書いたほうがわかりやすいかなと思いました．

``` python
nx1 = int(self.__psf_w * np.random.randn() + dps.x + 0.5)
nx2 = int(self.__psf_w * np.random.randn() + dps.y + 0.5)
if 0 <= nx1 < self.__naxis1 and 0 <= nx2 < self.__naxis2:  # 属性の方も名称を変更しています
    a[naxis2][naxis1] += 1
```

yamadayy replied:
同上

astronasutarou reviewed (changes requested):
……長らくお待たせいたしました！一通り目を通したつもりです．同じコメントが複数あったり言っていることが食い違ったりしているかもしれません……変なことを言っていたらご指摘ください．

細かな typo など，すぐに対応できそうな箇所を直していただければと思います．おそらく最近クラス名の変更があったのではないかと思いますが，コメントや関数名が追従していないような箇所がいくつか見受けられました．

また，現在の実装では `DetectorImage` はフルサイズの画像という理解でいいでしょうか？ simulation の最終生成物 (analysis の始点) は 9×9 の切り出し画像であるべきではないでしょうか？

情報を保持するクラスの大まかな構成はこれでいいと思いますが，細かい部分はいろいろとアップデートが必要だと感じました．また，現在の実装では各 builder にシミュレーションと解析両方の関数が実装されていますが，実際には大きく異なる問題を扱うのでベースとなるクラス (共有情報) を継承しつつ，別のクラスとして実装したほうがいいのではないかと思います．

個人的な感想ですが for-loop 等でスコープが限られているところはともかく，関数の引数などで 1,2 文字の変数を使われるとレビューする側として結構厳しいなと感じました．

YoshiyukiYamada committed:
Names of variables are fixed to adapt corresponding class name. Short variable name is fixed to appropreate name.

YoshiyukiYamada committed:
Names of variables are fixed to adapt corresponding class name. Short variable name is fixed to appropreate name.

YoshiyukiYamada committed:
Fix code climate error

yamadayy commented:
座標をFITSに合わせるかどうかは保留しています。magから光子数への変換はちゃんとした実装は保留していますが、関数としては変換関数を作りました(magをそのまま返すだけですが)。上の二つは、ちょっと設計の議論をさせてください。変数名は気が付く範囲で直しました。map_on_detector_builderのget_positions_on_detector関数に引数を増やす件は、codeclimateエラーが出ない範囲で対応しました。

YoshiyukiYamada committed:
Add poisson noise to _generate_a_stellar_image method.

YoshiyukiYamada committed:
Add poisson noise to _generate_a_stellar_image method.

YoshiyukiYamada committed:
Use to_value.

YoshiyukiYamada committed:
Apply review comments.

YoshiyukiYamada committed:
Apply review comments.

YoshiyukiYamada committed:
Apply codeclimate error.

astronasutarou commented on src/telescope_baseline/tools/pipeline/map_on_detector_builder.py:
```diff
@@ -0,0 +1,122 @@
+        if not (tmp[k][0] < 0 or tmp[k][1] < 0 or tmp[k][0] > self.__nx or tmp[k][1] > self.__ny):
+            a.append(PositionOnDetector(k, Position2D(tmp[k][0], tmp[k][1]), sky_positions[0].datetime, mag=3000))
+
+    def estimate_positions_on_detector(self, detector_image: DetectorImage, window_size: int, oversampling=4, maxiters=3)\
```
解決済み

YoshiyukiYamada committed:
Apply review comment.

yamadayy commented:
これまでにいただいたコメントは、すべて確認して、修正、あるいは保留などコメントしました。

astronasutarou commented:
僕の画面では Unresolved conversations が 9 件 (うち 4 件は対応いただいたものに対する追確認) 残っているのですが山田さんの方では確認できませんか？

YoshiyukiYamada committed:
Apply review comment.

yamadayy commented:
済みませんでした。見落としてました。保留も含めてコメントを返しました。

YoshiyukiYamada committed:
Apply review comment.

astronasutarou reviewed (approved):
コメントがすべて対応されたことを確認しました．ありがとうございます．
